src/mod_muc_db_mnesia.erl

```diff
@@ -0,0 +1,182 @@
+%%% published by the Free Software Foundation; either version 2 of the
+%%% License, or (at your option) any later version.
+%%%
+%%% This program is distributed in the hope that it will be useful,
+%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
+%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+%%% General Public License for more details.
+%%%
+%%% You should have received a copy of the GNU General Public License
+%%% along with this program; if not, write to the Free Software
+%%% Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+%%% 02111-1307 USA
+%%%
+%%%----------------------------------------------------------------------
+
+-module(mod_muc_db_mnesia).
+-behaviour(mod_muc_db).
+-export([init/2,
+         store_room/4,
+         restore_room/3,
+         forget_room/3,
+         get_rooms/2,
+         can_use_nick/4,
+         get_nick/3,
+         set_nick/4,
+         unset_nick/3]).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+-include("mod_muc.hrl").
+
+init(_Host, _Opts) ->
+    mnesia:create_table(muc_room,
+                        [{disc_copies, [node()]},
+                         {attributes, record_info(fields, muc_room)}]),
+    mnesia:create_table(muc_registered,
+                        [{disc_copies, [node()]},
+                         {attributes, record_info(fields, muc_registered)}]),
+    mnesia:add_table_copy(muc_room, node(), disc_copies),
+    mnesia:add_table_copy(muc_registered, node(), disc_copies),
+    mnesia:add_table_index(muc_registered, nick),
+    ok.
+
+-spec store_room(ejabberd:server(), ejabberd:server(), mod_muc:room(), list())
+            -> {'aborted', _} | {'atomic', _}.
+store_room(_ServerHost, MucHost, RoomName, Opts) ->
+    F = fun() ->
+                mnesia:write(#muc_room{name_host = {RoomName, MucHost},
+                                       opts = Opts})
+        end,
+    Result = mnesia:transaction(F),
+    case Result of
+        {atomic, _} ->
+            ok;
+        _ ->
+            ?ERROR_MSG("issue=store_room_failed room=~ts", [RoomName])
+    end,
+    Result.
+
+
+restore_room(_ServerHost, MucHost, RoomName) ->
+    try mnesia:dirty_read(muc_room, {RoomName, MucHost}) of
+        [#muc_room{opts = Opts}] ->
+            {ok, Opts};
+        [] ->
+            {error, room_not_found};
+        Other ->
+            {error, Other}
+        catch Class:Reason ->
+            ?ERROR_MSG("issue=restore_room_failed room=~ts reason=~p:~p",
+                       [RoomName, Class, Reason]),
+            {error, {Class, Reason}}
+    end.
+
+-spec forget_room(ejabberd:server(), ejabberd:server(), mod_muc:room()) -> 'ok'.
+forget_room(_ServerHost, MucHost, RoomName) ->
+    F = fun() ->
+                mnesia:delete({muc_room, {RoomName, MucHost}})
+        end,
+    Result = mnesia:transaction(F),
+    case Result of
+        {atomic, _} ->
+            ejabberd_hooks:run(forget_room, MucHost, [MucHost, RoomName]),
```

Comment:
A backend does not seem to be a good place to run hooks from; every backend will have to duplicate that line on success in `forget_room/3`.